File: backend/src/main/java/iuh/labbooking/service/device/DeviceService.java
package iuh.labbooking.service.device;

import iuh.labbooking.dto.request.device.BulkDeviceStatusRequest;
import iuh.labbooking.dto.request.device.DeviceRequest;
import iuh.labbooking.dto.response.base.PageResponse;
import iuh.labbooking.dto.response.device.DeviceAvailabilityResponse;
import iuh.labbooking.dto.response.device.DeviceResponse;
import iuh.labbooking.dto.response.device.SecureDeviceResponse;

import java.time.LocalDate;
import java.util.List;

public interface DeviceService {
    SecureDeviceResponse createDevice(DeviceRequest request);

    SecureDeviceResponse updateDevice(Long id, DeviceRequest request);

    SecureDeviceResponse findDeviceById(Long id);

    PageResponse<List<DeviceResponse>> findDevices(int page, int limit, String keyword);

    PageResponse<List<SecureDeviceResponse>> findAllDevicesForAdmin(int page, int limit, String keyword, Boolean active);

    void toggleDeviceStatus(Long id);

    void updateDevicesStatus(BulkDeviceStatusRequest request);


    List<DeviceAvailabilityResponse> findDeviceAvailability(
            Long labRoomId,
            LocalDate date,
            List<Long> slotIds,
            Long excludeBookingId);
}

File: frontend/src/services/device.service.ts
import http from '@/lib/http'
import type { ApiResponse, PageResponse } from '@/schemas/base.schema'
import type {
  DeviceRequest,
  DeviceResponse,
  DeviceAvailabilityParams,
  DeviceAvailabilityResponse,
  SecuredDeviceResponse,
  FindDevicesParams
} from '@/schemas/device.schema'

const deviceService = {
  create: (body: DeviceRequest) => http.post<ApiResponse<DeviceResponse>>('/devices', body),

  update: (id: number, body: DeviceRequest) => http.put<ApiResponse<DeviceResponse>>(`/devices/${id}`, body),

  getById: (id: number) => http.get<ApiResponse<DeviceResponse>>(`/devices/${id}`),

  filterDevices: (params: FindDevicesParams) =>
    http.get<ApiResponse<PageResponse<SecuredDeviceResponse[]>>>('/devices/admin', {
      params: {
        page: params.page || 1,
        limit: params.limit || 10,
        keyword: params.keyword,
        active: params.active
      }
    }),

  toggleActive: (id: number) => http.patch<ApiResponse<void>>(`/devices/${id}/status`),

  updateBulkStatus: (ids: number[], active: boolean) =>
    http.patch<ApiResponse<void>>('/devices/actions/bulk-active', { ids, active }),

  getAvailability: (params: DeviceAvailabilityParams) =>
    http.get<ApiResponse<DeviceAvailabilityResponse[]>>('/devices/availability', {
      params: {
        labRoomId: params.labRoomId,
        date: params.date,
        slotIds: params.slotIds.join(','),
        excludeBookingId: params.excludeBookingId
      }
    })
}

export default deviceService
